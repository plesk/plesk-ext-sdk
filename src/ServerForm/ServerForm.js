// Copyright 1999-2018. Plesk International GmbH. All rights reserved.

import { Component, Form, PropTypes } from '@plesk/ui-library';
import { withRouter } from 'react-router-dom';
import { withApi } from '../Api';
import { withConfig } from '../Config';
import { withStatusMessages } from '../StatusMessages';

export class ServerForm extends Component {
    static propTypes = {
        config: PropTypes.shape({
            baseUrl: PropTypes.string.isRequired,
        }).isRequired,
        action: PropTypes.string.isRequired,
        successUrl: PropTypes.string,
        successMessage: PropTypes.string,
        cancelUrl: PropTypes.string,
        children: PropTypes.node,
        history: PropTypes.shape({
            push: PropTypes.func.isRequired,
        }).isRequired,
        api: PropTypes.shape({
            post: PropTypes.func.isRequired,
        }).isRequired,
        statusMessages: PropTypes.shape({
            add: PropTypes.func.isRequired,
        }).isRequired,
        onSubmit: PropTypes.func,
        onSuccess: PropTypes.func,
        applyButton: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.object,
        ]),
        cancelButton: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.object,
        ]),
        staticContext: PropTypes.any,
    };

    static defaultProps = {
        successUrl: null,
        successMessage: null,
        cancelUrl: null,
        children: null,
        onSubmit: null,
        onSuccess: null,
        applyButton: false,
        cancelButton: false,
        staticContext: null,
    };

    state = {
        state: null,
        errors: {},
    };

    handleCancel = () => {
        const { cancelUrl, history, cancelButton } = this.props;

        if (cancelButton && typeof cancelButton.onClick === 'function' && cancelButton.onClick() === false) {
            return;
        }

        if (cancelUrl) {
            history.push(cancelUrl);
        }
    };

    handleSubmit = values => {
        const { config, action, onSubmit, api, statusMessages } = this.props;

        if (typeof onSubmit === 'function' && onSubmit(values) === false) {
            return;
        }

        this.setState({
            state: 'submit',
        });

        api.post(`${config.baseUrl}${action}`, values).then(({ status, errors, data }) => {
            const newState = {
                state: null,
            };

            if (Array.isArray(errors)) {
                errors.forEach(message => statusMessages.add({ intent: 'danger', message }));
            } else {
                newState.errors = errors;
            }

            this.setState(newState, () => {
                if (status === 'success') {
                    const { successUrl, successMessage, history, onSuccess } = this.props;

                    if (typeof onSuccess === 'function' && onSuccess(data) === false) {
                        return;
                    }

                    if (successUrl) {
                        history.push(successUrl);
                    }

                    if (successMessage) {
                        setTimeout(() => {
                            statusMessages.add({ intent: 'success', message: successMessage });
                        });
                    }
                }
            });
        });
    };

    prepareCancelButton() {
        const { cancelUrl, cancelButton } = this.props;

        if (cancelButton) {
            return {
                ...cancelButton,
                onClick: this.handleCancel,
            };
        }

        if (cancelUrl !== null) {
            return {
                onClick: this.handleCancel,
            };
        }

        return false;
    }

    render() {
        const { state, errors } = this.state;
        const { children, config, successUrl, successMessage, cancelUrl, history, onSubmit, cancelButton, onSuccess, api, statusMessages, staticContext, ...props } = this.props;

        return (
            <Form
                state={state}
                errors={errors}
                {...props}
                onSubmit={this.handleSubmit}
                cancelButton={this.prepareCancelButton()}
            >
                {children}
            </Form>
        );
    }
}

export default withStatusMessages(withRouter(withApi(withConfig(ServerForm))));
