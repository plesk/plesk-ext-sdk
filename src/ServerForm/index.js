// Copyright 1999-2025. WebPros International GmbH. All rights reserved.

export { default as ServerForm } from './ServerForm';
