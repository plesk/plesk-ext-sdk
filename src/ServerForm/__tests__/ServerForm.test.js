// Copyright 1999-2018. Plesk International GmbH. All rights reserved.
import { createElement } from 'react';
import { Form } from '@plesk/ui-library';
import { ServerForm } from '../ServerForm';

import { shallow } from 'enzyme';

jest.useFakeTimers();

describe('ServerForm', () => {
    let props;

    beforeEach(() => {
        props = {
            config: {
                baseUrl: '',
            },
            history: {
                push: jest.fn(),
            },
            api: {
                post: jest.fn(),
            },
            statusMessages: { add: jest.fn() },
            action: '/api/save',
        };
    });

    test('renders correctly', () => {
        const wrapper = shallow(
            <ServerForm
                {...props}
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('submits data with success', () => {
        const promise = Promise.resolve({ status: 'success' });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
                successUrl="/overview"
            />,
        );

        wrapper.find(Form).simulate('submit', { username: 'Bob' });

        expect.assertions(2);

        return promise.then(() => {
            expect(props.api.post).toHaveBeenCalledWith('/api/save', { username: 'Bob' });
            expect(props.history.push).toHaveBeenCalledWith('/overview');
        });
    });


    test('renders success message correctly', () => {
        const promise = Promise.resolve({ status: 'success' });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
                successMessage="Form saved"
            />,
        );

        wrapper.find(Form).simulate('submit');

        expect.assertions(1);

        return promise.then(() => {
            jest.runAllTimers();

            expect(props.statusMessages.add).toHaveBeenCalledWith({ intent: 'success', message: 'Form saved' });
        });
    });

    test('renders common error correctly', () => {
        const promise = Promise.resolve({ status: 'error', errors: ['Some error'] });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
            />,
        );

        wrapper.find(Form).simulate('submit');

        expect.assertions(1);

        return promise.then(() => {
            expect(props.statusMessages.add).toHaveBeenCalledWith({ intent: 'danger', message: 'Some error' });
        });
    });
});
