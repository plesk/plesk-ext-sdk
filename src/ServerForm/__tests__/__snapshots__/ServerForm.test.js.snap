// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ServerForm renders correctly 1`] = `
<Form
  action="/api/save"
  applyButton={false}
  cancelButton={false}
  errors={{}}
  onSubmit={[Function]}
  state={null}
/>
`;
